nix: "cannot find -llpeg: No such file or directory"

### Problem

Buliding on nix fails, probably because of the recent lpeg vendoring.

```
error: builder for '/nix/store/qhaj689hfcrzik1nzvichqlwvpdvhzzg-neovim-unwrapped-933fdff.drv' failed with exit code 2;
       last 10 log lines:
       > [  3%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/nlua0.c.o
       > [  3%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/rpc.c.o
       > -- Using NVIM_VERSION: v0.10.0-dev
       > [  3%] Built target update_version_stamp
       > [  3%] Linking C shared module ../../lib/libnlua0.so
       > /nix/store/5i11h3ji1cki45v3ndn1z6b3scma2jlm-binutils-2.40/bin/ld: cannot find -llpeg: No such file or directory
       > collect2: error: ld returned 1 exit status
       > make[2]: *** [src/nvim/CMakeFiles/nlua0.dir/build.make:178: lib/libnlua0.so] Error 1
       > make[1]: *** [CMakeFiles/Makefile2:562: src/nvim/CMakeFiles/nlua0.dir/all] Error 2
       > make: *** [Makefile:156: all] Error 2
       For full logs, run 'nix log /nix/store/qhaj689hfcrzik1nzvichqlwvpdvhzzg-neovim-unwrapped-933fdff.drv'.
error: 1 dependencies of derivation '/nix/store/al1m8db7xfw1vjrwf5gd0d8qvhvhhwr7-neovim-933fdff.drv' failed to build
error: 1 dependencies of derivation '/nix/store/y0h3pq8c5i09n7aimpcipp054f6m591n-home-manager-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/365wrdv763sylwqs28fygnhjryikszgb-man-paths.drv' failed to build
error: 1 dependencies of derivation '/nix/store/g2fqzprb9hqd8a68apjp8rs7h8n20dcz-neovim-933fdff-fish-completions.drv' failed to build
error: 1 dependencies of derivation '/nix/store/1cblr7wchrspgpa91kj82a12ziycpqk3-home-manager-generation.drv' failed to build
error: 1 dependencies of derivation '/nix/store/bq5chkj892p7m9v4hdf4r1ghylrnxfwi-user-environment.drv' failed to build
error: 1 dependencies of derivation '/nix/store/hq0xv5slsn46nbjq7y7w30xgbdskjnwp-etc.drv' failed to build
error: 1 dependencies of derivation '/nix/store/jnkyq7x1x74y3cvss1gdknpl540d2lc8-nixos-system-nixbox-23.05.20230428.937a9d1.drv' failed to build
```

### Steps to reproduce

Build using the default package in the latest `contrib/flake.nix`... the error should be guaranteed because of the nature of nix flakes.

### Expected behavior

Neovim should build using non-bundled lpeg.

### Neovim version (nvim -v)

NVIM v0.10.0-dev-9f0762f

### Vim (not Nvim) behaves the same?

?

### Operating system/version

Linux 6.2.12; NixOS

### Terminal name/version

foot 1.14.0

### $TERM environment variable

foot

### Installation

build from flake

The Nix build is definitely broken because it can't properly find lpeg while linking. If that `-llpeg` argument is instead manually replaced with the location of the lpeg library, the build will succeed, but this is a rather crude workaround that uses Ninja to build, and doesn't use the Lua in the Neovim package args. This is the workaround:

```patch
diff --git a/contrib/flake.nix b/contrib/flake.nix
index d38d1fd2b..ead8b0c78 100644
--- a/contrib/flake.nix
+++ b/contrib/flake.nix
@@ -19,6 +19,11 @@
           nativeBuildInputs = oa.nativeBuildInputs ++ [
             final.libiconv
           ];
+          configurePhase = ''
+            cmakeConfigurePhase
+            sed -i "s|-llpeg|${final.luajitPackages.lpeg}/lib/lua/5.1/lpeg.so|g" build.ninja
+          '';
+          buildInputs = oa.buildInputs ++ [final.ninja];
         });
 
         # a development binary to help debug issues
```

The problem is definitely in how lpeg is linked. I'm not too familiar with the CMake language, though, so I'm not of much help there. This is the full log for the failing Nix build, unmodified:

```
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/qndp94622fddpm3f5faw8zc1cyvk38wb-xrcvayf7g65aa1ii60gb8s34d9872nls-source
source root is xrcvayf7g65aa1ii60gb8s34d9872nls-source
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
applying patch /nix/store/jnp9n2x9cdvl47da3656illrd3p2qvb5-system_rplugin_manifest.patch
patching file runtime/autoload/remote/host.vim
patching file runtime/plugin/rplugin.vim
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
cmake flags: -DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=OFF -DCMA
KE_INSTALL_LOCALEDIR=/nix/store/jvyx1nlb171lw8a4cmm1334a2p25i009-neovim-unwrapped-c194acb/share/locale -DCMAKE_INSTALL_LIBEXECDIR=/nix/store/jvyx1nlb171lw8a4cmm1334a2p25i009-neovim-unwra
pped-c194acb/libexec -DCMAKE_INSTALL_LIBDIR=/nix/store/jvyx1nlb171lw8a4cmm1334a2p25i009-neovim-unwrapped-c194acb/lib -DCMAKE_INSTALL_DOCDIR=/nix/store/jvyx1nlb171lw8a4cmm1334a2p25i009-ne
ovim-unwrapped-c194acb/share/doc/nvim -DCMAKE_INSTALL_INFODIR=/nix/store/jvyx1nlb171lw8a4cmm1334a2p25i009-neovim-unwrapped-c194acb/share/info -DCMAKE_INSTALL_MANDIR=/nix/store/jvyx1nlb17
1lw8a4cmm1334a2p25i009-neovim-unwrapped-c194acb/share/man -DCMAKE_INSTALL_OLDINCLUDEDIR=/nix/store/jvyx1nlb171lw8a4cmm1334a2p25i009-neovim-unwrapped-c194acb/include -DCMAKE_INSTALL_INCLU
DEDIR=/nix/store/jvyx1nlb171lw8a4cmm1334a2p25i009-neovim-unwrapped-c194acb/include -DCMAKE_INSTALL_SBINDIR=/nix/store/jvyx1nlb171lw8a4cmm1334a2p25i009-neovim-unwrapped-c194acb/sbin -DCMA
KE_INSTALL_BINDIR=/nix/store/jvyx1nlb171lw8a4cmm1334a2p25i009-neovim-unwrapped-c194acb/bin -DCMAKE_INSTALL_NAME_DIR=/nix/store/jvyx1nlb171lw8a4cmm1334a2p25i009-neovim-unwrapped-c194acb/l
ib -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_OSX_SYSROOT= -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_STRIP=/nix/store/nlgyw2fv0cm8rkz8qm1jyw78vyif1bl9-gcc-wrapper-12.2.0/bin/strip -DCMAKE_RANL
IB=/nix/store/nlgyw2fv0cm8rkz8qm1jyw78vyif1bl9-gcc-wrapper-12.2.0/bin/ranlib -DCMAKE_AR=/nix/store/nlgyw2fv0cm8rkz8qm1jyw78vyif1bl9-gcc-wrapper-12.2.0/bin/ar -DCMAKE_C_COMPILER=gcc -DCMA
KE_CXX_COMPILER=g++ -DCMAKE_INSTALL_PREFIX=/nix/store/jvyx1nlb171lw8a4cmm1334a2p25i009-neovim-unwrapped-c194acb -DUSE_BUNDLED=OFF 
-- The C compiler identification is GNU 12.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /nix/store/nlgyw2fv0cm8rkz8qm1jyw78vyif1bl9-gcc-wrapper-12.2.0/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- CMAKE_INSTALL_PREFIX=/nix/store/jvyx1nlb171lw8a4cmm1334a2p25i009-neovim-unwrapped-c194acb
-- CMAKE_BUILD_TYPE=Release
-- Using Lua interpreter: /nix/store/dvjryghifiz80jy36a0kzhly679gvgp8-luajit-2.1.0-2022-10-04-env/bin/luajit
-- Using Lua interpreter for code generation: /nix/store/dvjryghifiz80jy36a0kzhly679gvgp8-luajit-2.1.0-2022-10-04-env/bin/luajit
-- Using Lua compiler: /nix/store/dvjryghifiz80jy36a0kzhly679gvgp8-luajit-2.1.0-2022-10-04-env/bin/luajit -b -s %s -
-- Could NOT find libuv (missing: libuv_DIR)
-- Looking for dlopen in dl
-- Looking for dlopen in dl - found
-- Looking for kstat_lookup in kstat
-- Looking for kstat_lookup in kstat - not found
-- Looking for kvm_open in kvm
-- Looking for kvm_open in kvm - not found
-- Looking for gethostbyname in nsl
-- Looking for gethostbyname in nsl - not found
-- Looking for perfstat_cpu in perfstat
-- Looking for perfstat_cpu in perfstat - not found
-- Looking for clock_gettime in rt
-- Looking for clock_gettime in rt - found
-- Looking for sendfile in sendfile
-- Looking for sendfile in sendfile - not found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Found Libuv: /nix/store/wcdi96dabywjdc03kr844plhk8mcsk5s-libuv-1.44.2/lib/libuv.so (Required is at least version "1.28.0") 
-- Found Libluv: /nix/store/bq4phzpnp28zvrjin10wsz511vz5maw7-libluv-1.44.2-1/lib/libluv.so (Required is at least version "1.43.0") 
-- Found Iconv: /nix/store/c4rbjkvlwcicy0z0v85h4lbi6qxk2m31-glibc-iconv-2.37/include  
-- Found Lpeg: /nix/store/dvjryghifiz80jy36a0kzhly679gvgp8-luajit-2.1.0-2022-10-04-env/lib/lua/5.1/lpeg.so  
-- Found Libtermkey: /nix/store/sm73b2amwm68a46k5i6gynp32lf80pyi-libtermkey-0.22/lib/libtermkey.so (Required is at least version "0.22") 
-- Found Libvterm: /nix/store/r24cprc7dny0is31hziccic59nb1cn26-libvterm-neovim-0.3.1/include (found suitable version "0.3", minimum required is "0.3") 
-- Found Msgpack: /nix/store/zgh2xvp2dlqch85fig5nq8ijyz6kcp45-msgpack-3.3.0/lib/libmsgpackc.so (found suitable version "3.3.0", minimum required is "1.0.0") 
-- Found Treesitter: /nix/store/q4wa1icf99nrvcibmcgq2n2h2mf29dqh-tree-sitter-0.20.8/lib/libtree-sitter.so (Required is at least version "0.20.8") 
-- Found Unibilium: /nix/store/lbpsk9w1zl451yxjsxmpzdzj0vmwqakc-unibilium-2.1.1/include (Required is at least version "2.0") 
-- Performing Test HAVE_WORKING_LIBINTL
-- Performing Test HAVE_WORKING_LIBINTL - Success
-- Looking for _nl_msg_cat_cntr
-- Looking for _nl_msg_cat_cntr - found
-- Found Libintl:   
-- Found Luajit: /nix/store/dvjryghifiz80jy36a0kzhly679gvgp8-luajit-2.1.0-2022-10-04-env/lib/libluajit-5.1.so  
-- Performing Test HAVE_WIMPLICIT_FALLTHROUGH_FLAG
-- Performing Test HAVE_WIMPLICIT_FALLTHROUGH_FLAG - Success
-- Performing Test HAS_DIAG_COLOR_FLAG
-- Performing Test HAS_DIAG_COLOR_FLAG - Success
-- Performing Test HAS_FSTACK_PROTECTOR_STRONG_FLAG
-- Performing Test HAS_FSTACK_PROTECTOR_STRONG_FLAG - Success
-- Found Gettext: /nix/store/rvw19191b3w7z0q99r0qnl5gwp3jr8n2-gettext-0.21/bin/msgmerge (found version "0.21") 
-- Performing Test HAVE_EXECINFO_BACKTRACE
-- Performing Test HAVE_EXECINFO_BACKTRACE - Success
-- Performing Test HAVE_BUILTIN_ADD_OVERFLOW
-- Performing Test HAVE_BUILTIN_ADD_OVERFLOW - Success
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of int
-- Check size of int - done
-- Check size of long
-- Check size of long - done
-- Check size of intmax_t
-- Check size of intmax_t - done
-- Check size of size_t
-- Check size of size_t - done
-- Check size of void *
-- Check size of void * - done
-- Looking for _NSGetEnviron
-- Looking for _NSGetEnviron - not found
-- Looking for include file langinfo.h
-- Looking for include file langinfo.h - found
-- Looking for include file strings.h
-- Looking for include file strings.h - found
-- Looking for include file sys/utsname.h
-- Looking for include file sys/utsname.h - found
-- Looking for include file termios.h
-- Looking for include file termios.h - found
-- Looking for include file sys/uio.h
-- Looking for include file sys/uio.h - found
-- Looking for include file sys/sdt.h
-- Looking for include file sys/sdt.h - not found
-- Looking for fseeko
-- Looking for fseeko - found
-- Looking for readv
-- Looking for readv - found
-- Looking for readlink
-- Looking for readlink - found
-- Looking for strnlen
-- Looking for strnlen - found
-- Looking for strcasecmp
-- Looking for strcasecmp - found
-- Looking for strncasecmp
-- Looking for strncasecmp - found
-- Looking for strptime
-- Looking for strptime - found
-- Performing Test HAVE_DIRFD_AND_FLOCK
-- Performing Test HAVE_DIRFD_AND_FLOCK - Success
-- Performing Test HAVE_PWD_FUNCS
-- Performing Test HAVE_PWD_FUNCS - Success
-- Looking for FD_CLOEXEC
-- Looking for FD_CLOEXEC - found
-- Looking for CODESET
-- Looking for CODESET - found
-- Looking for include file endian.h
-- Looking for include file endian.h - found
-- Performing Test HAVE_BE64TOH_MACROS
-- Performing Test HAVE_BE64TOH_MACROS - Success
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    BUILD_TESTING
    CMAKE_CXX_COMPILER
    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
    CMAKE_POLICY_DEFAULT_CMP0025
    USE_BUNDLED


-- Build files have been written to: /build/xrcvayf7g65aa1ii60gb8s34d9872nls-source/build
cmake: enabled parallel building
cmake: enabled parallel installing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
build flags: -j32 SHELL=/nix/store/zlf0f88vj30sc7567b80l52d19pbdmy2-bash-5.2-p15/bin/bash
[  0%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/lmpack.c.o
[  0%] Built target nvim_runtime_deps
[  0%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/conv.c.o
[  3%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/mpack_core.c.o
[  3%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/object.c.o
[  3%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/rpc.c.o
[  3%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/nlua0.c.o
-- Using NVIM_VERSION: v0.10.0-dev
[  3%] Built target update_version_stamp
[  3%] Linking C shared module ../../lib/libnlua0.so
/nix/store/f4qnwzv6y0nq8lix33jr5ykkyybs6fxf-binutils-2.40/bin/ld: cannot find -llpeg: No such file or directory
collect2: error: ld returned 1 exit status
make[2]: *** [src/nvim/CMakeFiles/nlua0.dir/build.make:178: lib/libnlua0.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:562: src/nvim/CMakeFiles/nlua0.dir/all] Error 2
make: *** [Makefile:156: all] Error 2
```

> > I can also reproduce this (in the nix shell provided by contrib/flake.nix) by calling:
> > So it's probably not a nix issue.
> 
> That doesn't make any sense.

```console
nix develop ./contrib
make USE_BUNDLED=OFF
mkdir -p build
touch build/.ran-deps-cmake
cd build && cmake -G 'Unix Makefiles' -DCMAKE_BUILD_TYPE=Debug  /home/mrcjk/git/github/forks/nvim/neovim/
-- The C compiler identification is Clang 14.0.6
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /nix/store/w2nhdkmich4l416ip426qmcw9jja5f8z-clang-wrapper-14.0.6/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- CMAKE_INSTALL_PREFIX=/var/empty/local
-- CMAKE_BUILD_TYPE=Debug
-- Using Lua interpreter: /nix/store/2g9349vhjq5mcrwlg9bmglnfsi1d19zh-luajit-2.1.0-2022-10-04/bin/luajit
-- Using Lua interpreter for code generation: /nix/store/2g9349vhjq5mcrwlg9bmglnfsi1d19zh-luajit-2.1.0-2022-10-04/bin/luajit
-- Using Lua compiler: /nix/store/2g9349vhjq5mcrwlg9bmglnfsi1d19zh-luajit-2.1.0-2022-10-04/bin/luajit -b -s %s -
-- Could NOT find libuv (missing: libuv_DIR)
-- Looking for dlopen in dl
-- Looking for dlopen in dl - found
-- Looking for kstat_lookup in kstat
-- Looking for kstat_lookup in kstat - not found
-- Looking for kvm_open in kvm
-- Looking for kvm_open in kvm - not found
-- Looking for gethostbyname in nsl
-- Looking for gethostbyname in nsl - not found
-- Looking for perfstat_cpu in perfstat
-- Looking for perfstat_cpu in perfstat - not found
-- Looking for clock_gettime in rt
-- Looking for clock_gettime in rt - found
-- Looking for sendfile in sendfile
-- Looking for sendfile in sendfile - not found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found Libuv: /nix/store/wcdi96dabywjdc03kr844plhk8mcsk5s-libuv-1.44.2/lib/libuv.so (Required is at least version "1.28.0")
-- Found Libluv: /nix/store/bq4phzpnp28zvrjin10wsz511vz5maw7-libluv-1.44.2-1/lib/libluv.so (Required is at least version "1.43.0")
-- Found Iconv: /nix/store/c4rbjkvlwcicy0z0v85h4lbi6qxk2m31-glibc-iconv-2.37/include
-- Found Lpeg: /nix/store/ry7wgv4ny3dh7nifdrjcl7m71b6d2a30-luajit2.1-lpeg-1.0.2-1/lib/lua/5.1/lpeg.so
-- Found Libtermkey: /nix/store/sm73b2amwm68a46k5i6gynp32lf80pyi-libtermkey-0.22/lib/libtermkey.so (Required is at least version "0.22")
-- Found Libvterm: /nix/store/r24cprc7dny0is31hziccic59nb1cn26-libvterm-neovim-0.3.1/include (found suitable version "0.3", minimum required is "0.3")
-- Found Msgpack: /nix/store/zgh2xvp2dlqch85fig5nq8ijyz6kcp45-msgpack-3.3.0/lib/libmsgpackc.so (found suitable version "3.3.0", minimum required is "1.0.0")
-- Found Treesitter: /nix/store/q4wa1icf99nrvcibmcgq2n2h2mf29dqh-tree-sitter-0.20.8/lib/libtree-sitter.so (Required is at least version "0.20.8")
-- Found Unibilium: /nix/store/lbpsk9w1zl451yxjsxmpzdzj0vmwqakc-unibilium-2.1.1/include (Required is at least version "2.0")
-- Performing Test HAVE_WORKING_LIBINTL
-- Performing Test HAVE_WORKING_LIBINTL - Success
-- Looking for _nl_msg_cat_cntr
-- Looking for _nl_msg_cat_cntr - found
-- Found Libintl:
-- Found Luajit: /nix/store/2g9349vhjq5mcrwlg9bmglnfsi1d19zh-luajit-2.1.0-2022-10-04/lib/libluajit-5.1.so
-- Performing Test HAVE_WIMPLICIT_FALLTHROUGH_FLAG
-- Performing Test HAVE_WIMPLICIT_FALLTHROUGH_FLAG - Success
-- Performing Test HAS_DIAG_COLOR_FLAG
-- Performing Test HAS_DIAG_COLOR_FLAG - Success
-- Performing Test HAS_FSTACK_PROTECTOR_STRONG_FLAG
-- Performing Test HAS_FSTACK_PROTECTOR_STRONG_FLAG - Success
-- Found Gettext: /nix/store/rvw19191b3w7z0q99r0qnl5gwp3jr8n2-gettext-0.21/bin/msgmerge (found version "0.21")
-- Performing Test HAVE_EXECINFO_BACKTRACE
-- Performing Test HAVE_EXECINFO_BACKTRACE - Success
-- Performing Test HAVE_BUILTIN_ADD_OVERFLOW
-- Performing Test HAVE_BUILTIN_ADD_OVERFLOW - Success
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of int
-- Check size of int - done
-- Check size of long
-- Check size of long - done
-- Check size of intmax_t
-- Check size of intmax_t - done
-- Check size of size_t
-- Check size of size_t - done
-- Check size of void *
-- Check size of void * - done
-- Looking for _NSGetEnviron
-- Looking for _NSGetEnviron - not found
-- Looking for include file langinfo.h
-- Looking for include file langinfo.h - found
-- Looking for include file strings.h
-- Looking for include file strings.h - found
-- Looking for include file sys/utsname.h
-- Looking for include file sys/utsname.h - found
-- Looking for include file termios.h
-- Looking for include file termios.h - found
-- Looking for include file sys/uio.h
-- Looking for include file sys/uio.h - found
-- Looking for include file sys/sdt.h
-- Looking for include file sys/sdt.h - not found
-- Looking for fseeko
-- Looking for fseeko - found
-- Looking for readv
-- Looking for readv - found
-- Looking for readlink
-- Looking for readlink - found
-- Looking for strnlen
-- Looking for strnlen - found
-- Looking for strcasecmp
-- Looking for strcasecmp - found
-- Looking for strncasecmp
-- Looking for strncasecmp - found
-- Looking for strptime
-- Looking for strptime - found
-- Performing Test HAVE_DIRFD_AND_FLOCK
-- Performing Test HAVE_DIRFD_AND_FLOCK - Success
-- Performing Test HAVE_PWD_FUNCS
-- Performing Test HAVE_PWD_FUNCS - Success
-- Looking for FD_CLOEXEC
-- Looking for FD_CLOEXEC - found
-- Looking for CODESET
-- Looking for CODESET - found
-- Looking for include file endian.h
-- Looking for include file endian.h - found
-- Performing Test HAVE_BE64TOH_MACROS
-- Performing Test HAVE_BE64TOH_MACROS - Success
-- Configuring done
-- Generating done
-- Build files have been written to: /home/mrcjk/git/github/forks/nvim/neovim/build
touch build/.ran-cmake
make -C build
make[1]: Entering directory '/home/mrcjk/git/github/forks/nvim/neovim/build'
-- Using NVIM_VERSION: v0.10.0-dev-7e3d6ff40-dirty
[  0%] Built target update_version_stamp
[  0%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/conv.c.o
[  0%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/lmpack.c.o
[  0%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/mpack_core.c.o
[  0%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/object.c.o
[  0%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/rpc.c.o
[  4%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/nlua0.c.o
[  4%] Linking C shared module ../../lib/libnlua0.so
/nix/store/f4qnwzv6y0nq8lix33jr5ykkyybs6fxf-binutils-2.40/bin/ld: cannot find -llpeg: No such file or directory
clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [src/nvim/CMakeFiles/nlua0.dir/build.make:178: lib/libnlua0.so] Error 1
make[2]: *** [CMakeFiles/Makefile2:562: src/nvim/CMakeFiles/nlua0.dir/all] Error 2
make[1]: *** [Makefile:156: all] Error 2
make[1]: Leaving directory '/home/mrcjk/git/github/forks/nvim/neovim/build'
make: *** [Makefile:84: nvim] Error 2
```

Could also reproduce on archlinux:

<details>
<summary>Build Log</summary>

```
❯ mkdir build
❯ cd build
❯ cmake ..
-- The C compiler identification is GNU 12.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- CMAKE_INSTALL_PREFIX=/usr/local
-- CMAKE_BUILD_TYPE not specified, default is 'Debug'
-- Using Lua interpreter: /usr/bin/luajit
-- Using Lua interpreter for code generation: /usr/bin/luajit
-- Using Lua compiler: /usr/bin/luajit -b -s %s -
-- Could NOT find libuv (missing: libuv_DIR)
-- Looking for dlopen in dl
-- Looking for dlopen in dl - found
-- Looking for kstat_lookup in kstat
-- Looking for kstat_lookup in kstat - not found
-- Looking for kvm_open in kvm
-- Looking for kvm_open in kvm - not found
-- Looking for gethostbyname in nsl
-- Looking for gethostbyname in nsl - found
-- Looking for perfstat_cpu in perfstat
-- Looking for perfstat_cpu in perfstat - not found
-- Looking for clock_gettime in rt
-- Looking for clock_gettime in rt - found
-- Looking for sendfile in sendfile
-- Looking for sendfile in sendfile - not found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Found Libuv: /usr/lib/libuv.so (Required is at least version "1.28.0") 
-- Found Libluv: /usr/lib/libluv.so (Required is at least version "1.43.0") 
-- Found Iconv: /usr/include  
-- Found Lpeg: /usr/lib/lua/5.1/lpeg.so  
-- Found Libtermkey: /usr/lib/libtermkey.so (Required is at least version "0.22") 
-- Found Libvterm: /usr/include (found suitable version "0.3", minimum required is "0.3") 
-- Found Msgpack: /usr/lib/libmsgpackc.so (found suitable version "5.0.0", minimum required is "1.0.0") 
-- Found Treesitter: /usr/lib/libtree-sitter.so (Required is at least version "0.20.8") 
-- Found Unibilium: /usr/include (Required is at least version "2.0") 
-- Performing Test HAVE_WORKING_LIBINTL
-- Performing Test HAVE_WORKING_LIBINTL - Success
-- Looking for _nl_msg_cat_cntr
-- Looking for _nl_msg_cat_cntr - found
-- Found Libintl:   
-- Found Luajit: /usr/lib/libluajit-5.1.so  
-- Performing Test HAVE_WIMPLICIT_FALLTHROUGH_FLAG
-- Performing Test HAVE_WIMPLICIT_FALLTHROUGH_FLAG - Success
-- Performing Test HAS_DIAG_COLOR_FLAG
-- Performing Test HAS_DIAG_COLOR_FLAG - Success
-- Performing Test HAS_FSTACK_PROTECTOR_STRONG_FLAG
-- Performing Test HAS_FSTACK_PROTECTOR_STRONG_FLAG - Success
-- Found Gettext: /usr/bin/msgmerge (found version "0.21.1") 
-- Performing Test HAVE_EXECINFO_BACKTRACE
-- Performing Test HAVE_EXECINFO_BACKTRACE - Success
-- Performing Test HAVE_BUILTIN_ADD_OVERFLOW
-- Performing Test HAVE_BUILTIN_ADD_OVERFLOW - Success
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of int
-- Check size of int - done
-- Check size of long
-- Check size of long - done
-- Check size of intmax_t
-- Check size of intmax_t - done
-- Check size of size_t
-- Check size of size_t - done
-- Check size of void *
-- Check size of void * - done
-- Looking for _NSGetEnviron
-- Looking for _NSGetEnviron - not found
-- Looking for include file langinfo.h
-- Looking for include file langinfo.h - found
-- Looking for include file strings.h
-- Looking for include file strings.h - found
-- Looking for include file sys/utsname.h
-- Looking for include file sys/utsname.h - found
-- Looking for include file termios.h
-- Looking for include file termios.h - found
-- Looking for include file sys/uio.h
-- Looking for include file sys/uio.h - found
-- Looking for include file sys/sdt.h
-- Looking for include file sys/sdt.h - found
-- Looking for fseeko
-- Looking for fseeko - found
-- Looking for readv
-- Looking for readv - found
-- Looking for readlink
-- Looking for readlink - found
-- Looking for strnlen
-- Looking for strnlen - found
-- Looking for strcasecmp
-- Looking for strcasecmp - found
-- Looking for strncasecmp
-- Looking for strncasecmp - found
-- Looking for strptime
-- Looking for strptime - found
-- Performing Test HAVE_DIRFD_AND_FLOCK
-- Performing Test HAVE_DIRFD_AND_FLOCK - Success
-- Performing Test HAVE_PWD_FUNCS
-- Performing Test HAVE_PWD_FUNCS - Success
-- Looking for FD_CLOEXEC
-- Looking for FD_CLOEXEC - found
-- Looking for CODESET
-- Looking for CODESET - found
-- Looking for include file endian.h
-- Looking for include file endian.h - found
-- Performing Test HAVE_BE64TOH_MACROS
-- Performing Test HAVE_BE64TOH_MACROS - Success
-- Configuring done
-- Generating done
-- Build files have been written to: /home/traxys/Softs/neovim/build
cmake ..  5.32s user 3.07s system 98% cpu 8.478 total
❯ make -j1
-- Using NVIM_VERSION: v0.10.0-dev-233+g066db5e7c
[  0%] Built target update_version_stamp
[  0%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/conv.c.o
[  0%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/lmpack.c.o
[  0%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/mpack_core.c.o
[  0%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/object.c.o
[  0%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/mpack/rpc.c.o
[  4%] Building C object src/nvim/CMakeFiles/nlua0.dir/__/nlua0.c.o
[  4%] Linking C shared module ../../lib/libnlua0.so
/usr/bin/ld: cannot find -llpeg: No such file or directory
collect2: error: ld returned 1 exit status
make[2]: *** [src/nvim/CMakeFiles/nlua0.dir/build.make:178: lib/libnlua0.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:562: src/nvim/CMakeFiles/nlua0.dir/all] Error 2
make: *** [Makefile:156: all] Error 2
```

</details>

No sorry I was not aware

<details>
<summary>Build log with toplevel make USE_BUNDLED=OFF</summary>

```
❯ make USE_BUNDLED=OFF
mkdir -p build
touch build/.ran-deps-cmake
cd build && cmake -G 'Ninja' -DCMAKE_BUILD_TYPE=Debug  /home/traxys/Softs/neovim/
-- The C compiler identification is GNU 12.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- CMAKE_INSTALL_PREFIX=/usr/local
-- CMAKE_BUILD_TYPE=Debug
-- Using Lua interpreter: /usr/bin/luajit
-- Using Lua interpreter for code generation: /usr/bin/luajit
-- Using Lua compiler: /usr/bin/luajit -b -s %s -
-- Found Libluv: /usr/lib/libluv.so (Required is at least version "1.43.0") 
-- Found Iconv: /usr/include  
-- Found Lpeg: /usr/lib/lua/5.1/lpeg.so  
-- Found Libtermkey: /home/traxys/Softs/neovim/.deps/usr/lib/libtermkey.a (Required is at least version "0.22") 
-- Found Libvterm: /home/traxys/Softs/neovim/.deps/usr/include (found suitable version "0.3", minimum required is "0.3") 
-- Found Msgpack: /home/traxys/Softs/neovim/.deps/usr/lib/libmsgpack-c.a (found suitable version "6.0.0", minimum required is "1.0.0") 
-- Found Treesitter: /home/traxys/Softs/neovim/.deps/usr/lib/libtree-sitter.a (Required is at least version "0.20.8") 
-- Found Unibilium: /home/traxys/Softs/neovim/.deps/usr/include (Required is at least version "2.0") 
-- Performing Test HAVE_WORKING_LIBINTL
-- Performing Test HAVE_WORKING_LIBINTL - Success
-- Looking for _nl_msg_cat_cntr
-- Looking for _nl_msg_cat_cntr - found
-- Found Libintl:   
-- Found Luajit: /usr/lib/libluajit-5.1.so  
-- Performing Test HAVE_WIMPLICIT_FALLTHROUGH_FLAG
-- Performing Test HAVE_WIMPLICIT_FALLTHROUGH_FLAG - Success
-- Performing Test HAS_DIAG_COLOR_FLAG
-- Performing Test HAS_DIAG_COLOR_FLAG - Success
-- Performing Test HAS_FSTACK_PROTECTOR_STRONG_FLAG
-- Performing Test HAS_FSTACK_PROTECTOR_STRONG_FLAG - Success
-- Found Gettext: /usr/bin/msgmerge (found version "0.21.1") 
-- Performing Test HAVE_EXECINFO_BACKTRACE
-- Performing Test HAVE_EXECINFO_BACKTRACE - Success
-- Performing Test HAVE_BUILTIN_ADD_OVERFLOW
-- Performing Test HAVE_BUILTIN_ADD_OVERFLOW - Success
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of int
-- Check size of int - done
-- Check size of long
-- Check size of long - done
-- Check size of intmax_t
-- Check size of intmax_t - done
-- Check size of size_t
-- Check size of size_t - done
-- Check size of void *
-- Check size of void * - done
-- Looking for _NSGetEnviron
-- Looking for _NSGetEnviron - not found
-- Looking for include file langinfo.h
-- Looking for include file langinfo.h - found
-- Looking for include file strings.h
-- Looking for include file strings.h - found
-- Looking for include file sys/utsname.h
-- Looking for include file sys/utsname.h - found
-- Looking for include file termios.h
-- Looking for include file termios.h - found
-- Looking for include file sys/uio.h
-- Looking for include file sys/uio.h - found
-- Looking for include file sys/sdt.h
-- Looking for include file sys/sdt.h - found
-- Looking for fseeko
-- Looking for fseeko - found
-- Looking for readv
-- Looking for readv - found
-- Looking for readlink
-- Looking for readlink - found
-- Looking for strnlen
-- Looking for strnlen - found
-- Looking for strcasecmp
-- Looking for strcasecmp - found
-- Looking for strncasecmp
-- Looking for strncasecmp - found
-- Looking for strptime
-- Looking for strptime - found
-- Performing Test HAVE_DIRFD_AND_FLOCK
-- Performing Test HAVE_DIRFD_AND_FLOCK - Success
-- Performing Test HAVE_PWD_FUNCS
-- Performing Test HAVE_PWD_FUNCS - Success
-- Looking for FD_CLOEXEC
-- Looking for FD_CLOEXEC - found
-- Looking for CODESET
-- Looking for CODESET - found
-- Looking for include file endian.h
-- Looking for include file endian.h - found
-- Performing Test HAVE_BE64TOH_MACROS
-- Performing Test HAVE_BE64TOH_MACROS - Success
-- Configuring done
-- Generating done
-- Build files have been written to: /home/traxys/Softs/neovim/build
touch build/.ran-cmake
ninja  -C build
ninja: Entering directory `build'
[0/2] Re-checking globbed directories...
[2/404] cd /home/traxys/Softs/neovim/build/src/nvim && /us...m -P /home/traxys/Softs/neovim/cmake/GenerateVersion.cmak
-- Using NVIM_VERSION: v0.10.0-dev-233+g066db5e7c
[8/404] Linking C shared module lib/libnlua0.so
FAILED: lib/libnlua0.so 
: && /usr/bin/cc -fPIC -g   -shared  -o lib/libnlua0.so src/nvim/CMakeFiles/nlua0.dir/__/mpack/conv.c.o src/nvim/CMakeFiles/nlua0.dir/__/mpack/lmpack.c.o src/nvim/CMakeFiles/nlua0.dir/__/mpack/mpack_core.c.o src/nvim/CMakeFiles/nlua0.dir/__/mpack/object.c.o src/nvim/CMakeFiles/nlua0.dir/__/mpack/rpc.c.o src/nvim/CMakeFiles/nlua0.dir/__/nlua0.c.o -L/usr/lib/lua/5.1 -Wl,-rpath,/usr/lib/lua/5.1  -llpeg && :
/usr/bin/ld: cannot find -llpeg: No such file or directory
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
make: *** [Makefile:84: nvim] Error 1
make USE_BUNDLED=OFF  4.57s user 2.11s system 88% cpu 7.525 total
```

</details>

--- a/cmake/FindLpeg.cmake
+++ b/cmake/FindLpeg.cmake
@@ -10,5 +10,7 @@ ubuntu()
 find_package_handle_standard_args(Lpeg DEFAULT_MSG LPEG_LIBRARY)
 mark_as_advanced(LPEG_LIBRARY)
 
-add_library(lpeg INTERFACE)
-target_link_libraries(lpeg INTERFACE ${LPEG_LIBRARY})
+# Workaround: use an imported library to prevent cmake from modifying library
+# link path. See #23395.
+add_library(lpeg UNKNOWN IMPORTED)
+set_target_properties(lpeg PROPERTIES IMPORTED_LOCATION ${LPEG_LIBRARY})

